### PR TITLE
Added Matter bridge application

### PR DESCRIPTION
**Pull request by kkasperczyk-no**

### CODE REVIEW

**Review by wiba-nordic — CHANGES_REQUESTED** [review]
Reviewing the documentation in place of Greg.

I noticed there was a TODO section, so I guess this is still a work in progress, but I looked through the sections that looked finished.

File: applications/matter_bridge/README.rst
Comment (kkasperczyk-no) [reply applications/matter_bridge/README.rst:202]: It's a copy paste error. There should be only bridge node ID.

**Comment by kkasperczyk-no** [comment]
> Reviewing the documentation in place of Greg.
> 
> I noticed there was a TODO section, so I guess this is still a work in progress, but I looked through the sections that looked finished.

Thanks @wiba-nordic! I addressed your suggestions and filled missing TODO section (see last commit).

File: applications/matter_bridge/Kconfig
Comment (Damian-Nordic) [inline applications/matter_bridge/Kconfig:44]: I think other apps already switched to using ZEPHYR_CONNECTEDHOMEIP_MODULE_DIR directly.

File: applications/matter_bridge/README.rst
Comment (wiba-nordic) [inline applications/matter_bridge/README.rst:277]:
```suggestion
For this application, you can use one of the following :ref:`onboarding information formats <ug_matter_network_topologies_commissioning_onboarding_formats>` to provide the commissioner with the data payload that includes the device discriminator and the setup PIN code:
```

**Review by wiba-nordic — CHANGES_REQUESTED** [review]
One very small change, and I just realized there's no release note entry. Seeing as it's a new application, one is warranted.

File: applications/matter_bridge/README.rst
Comment (greg-fer) [inline applications/matter_bridge/README.rst:108]:
```suggestion
.. code-block:: console
```

File: applications/matter_bridge/README.rst
Comment (greg-fer) [inline applications/matter_bridge/README.rst:132]:
```suggestion
.. code-block:: console
```

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst
Comment (wiba-nordic) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:234]:
```suggestion
* Added the :ref:`Matter bridge <matter_bridge_app>` application.
```

File: applications/matter_bridge/README.rst
Comment (greg-fer) [inline applications/matter_bridge/README.rst:107]: In the doc output, it will be hard to distinguish information about the two commands. I suggest you add lines 108-130 in a bullet point, and lines 132-146 in another. You can have:
```suggestion
You can use the following commands to control the bridge device:

* For adding a new bridged device to the bridge:

...

* For removes a new bridged device from the bridge:

```

File: applications/matter_bridge/README.rst
Comment (Damian-Nordic) [inline applications/matter_bridge/README.rst:130]: Add a line that displays the assigned endpoint ID so that it's clear how a user knows the endpoint ID for the "remove" command?

**Review by Damian-Nordic — APPROVED** [review]

File: applications/matter_bridge/README.rst
Comment (greg-fer) [inline applications/matter_bridge/README.rst:136]: I would drop "new" here, because this would suggest you can only remove the last added bridge.

File: applications/matter_bridge/README.rst
Comment (greg-fer) [inline applications/matter_bridge/README.rst:139]:
```suggestion
   * ``<bridged_device_endpoint_id>`` is the endpoint ID of the bridged device to be removed.
```

**Review by greg-fer — COMMENTED** [review]
Couple of suggestions.

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst
Comment (wiba-nordic) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:359]:
```suggestion
  * Added the :ref:`Matter thermostat <matter_thermostat_sample>` sample.
```
This is existing text, but might as well fix it since we're editing a close-by line anyway.

**Review by wiba-nordic — CHANGES_REQUESTED** [review]

File: applications/matter_bridge/README.rst
Comment (kkasperczyk-no) [reply applications/matter_bridge/README.rst:107]: Ok, I think it was like this before the review from @wiba-nordic, so can we have some agreement on that?

Comment (kkasperczyk-no) [reply applications/matter_bridge/README.rst:107]: Or maybe not, it wasn't the same. Ok I will introduce it :)

Comment (kkasperczyk-no) [reply applications/matter_bridge/README.rst:107]: Ok, done, I hope it is correct now.

**Review by wiba-nordic — COMMENTED** [review]
Looks good to my eye, but I'll check the doc build before giving approval.

File: applications/matter_bridge/README.rst
Comment (wiba-nordic) [inline applications/matter_bridge/README.rst:118]:
```suggestion
     The argument is mandatory and accepts the following values:
```
Wrong indent breaks doc build.

**Review by markaj-nordic — APPROVED** [review]
The implementation was discussed f2f prior to this PR, LGTM.

**Review by ArekBalysNordic — APPROVED** [review]

**Comment by wiba-nordic** [comment]
We need 
```
   ../../applications/matter_bridge/README
```
as the second entry in `applications.rst`.

File: samples/matter/common/src/bridge/bridged_device_data_provider.h
Comment (markaj-nordic) [inline samples/matter/common/src/bridge/bridged_device_data_provider.h:16]: I realized we missed virtual destructor here resulting in memory leaks when destroying child objects. I can fix it the follow up PR however.

**Review by greg-fer — CHANGES_REQUESTED** [review]
We also need to:
- Mention this new application in all lines that say to check out the Matter weather station as the only Matter application, for example here:
image
Check for `matter_weather_station_app` and update the following pages (these can just be updated with :ref:`Matter applications<applications>`): protocols/matter/index.rst, matter/tools.rst, samples/matter.rst
- Update hw_requirements.rst with the memory values for the new application.
- Update data_model.rst with entries for the new application (and thermostat sample).

File: applications/matter_bridge/README.rst
Comment (wiba-nordic) [inline applications/matter_bridge/README.rst:142]:
```suggestion
* For removing a bridged device from the bridge:
```

File: doc/nrf/applications.rst
Comment (wiba-nordic) [inline doc/nrf/applications.rst:26]: Should be between connectivity bridge and asset track, so that they are in alphabetical order based on the page title.
image

File: samples/matter/common/src/bridge/bridged_device_data_provider.h
Comment (ArekBalysNordic) [reply samples/matter/common/src/bridge/bridged_device_data_provider.h:16]: Please fix it in your PR, since we should finish this PR to unblock you.

**Review by wiba-nordic — CHANGES_REQUESTED** [review]

**Comment by ArekBalysNordic** [comment]
@wiba-nordic @greg-fer I've provided all changes. We decided to postpone generating a memory report until we have the final support for the bridge. Could you please review the last commit?

**Review by wiba-nordic — APPROVED** [review]
Everything looks good to me.

File: applications/matter_bridge/src/bridged_device_types/onoff_light_data_provider.cpp
Comment (markaj-nordic) [inline applications/matter_bridge/src/bridged_device_types/onoff_light_data_provider.cpp:42]: Not a blocker for this PR but should we really call `NotifyUpdateState()` here? To me, the the current flow is as follows:
- the Matter attribute is altered
- we call `BridgeManager::HandleWrite` which results in updating the bridged device state AND synchronizing the data provider as well by issuing polymorphic `BridgedDeviceDataProvider::UpdateState`
- in `OnOffLightDataProvider::UpdateState` we call `NotifyUpdateState` which brings us back to `BridgeManager::HandleUpdate` resulting in Matter bridged device attribute update which essentially was a trigger of the change in the first place

Comment (ArekBalysNordic) [reply applications/matter_bridge/src/bridged_device_types/onoff_light_data_provider.cpp:42]: Do you suggest repairing it here or unblocking your PR and considering it as a follow-up?

Comment (markaj-nordic) [reply applications/matter_bridge/src/bridged_device_types/onoff_light_data_provider.cpp:42]: I think we need @kkasperczyk-no 's insight here. For me we can live with it for now and fix later if needed.

File: doc/nrf/protocols/matter/overview/data_model.rst
Comment (greg-fer) [inline doc/nrf/protocols/matter/overview/data_model.rst:307]:
```suggestion
.. _ug_matter_device_types_utility:
```

File: applications/matter_bridge/README.rst
Comment (greg-fer) [inline applications/matter_bridge/README.rst:4]:
```suggestion
Matter bridge
##############
```
I would treat this on-par with "Matter controller" and other network topology devices.
See https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/protocols/matter/overview/network_topologies.html#matter-network-concepts.
Also because, assuming that "Matter:" is here to indicate the protocol and the second part is the device type, this could be interpreted we're adding an application for controlling bridges. ;)

**Review by greg-fer — CHANGES_REQUESTED** [review]
Really sorry, two more changes.

**Review by greg-fer — APPROVED** [review]
Ship it! 🚀

File: applications/matter_bridge/src/bridged_device_types/onoff_light_data_provider.cpp
Comment (ArekBalysNordic) [reply applications/matter_bridge/src/bridged_device_types/onoff_light_data_provider.cpp:42]: Ok, leave it as it is currently, and make a follow-up since it is the base of the whole project. :)

**Pull request merged**
